### PR TITLE
Auto capture

### PR DESCRIPTION
**Pull request by mayankamencherla**
Integrated auto capture of orders API into the magento razorpay extension.

### CODE REVIEW

File: package.xml [hunk 1]
```diff
@@ -11,33 +11,35 @@
     <notes>This version uses the Razorpay API v1 for the best possible experience.</notes>
     <authors>
         <author>
-            <name>Vivek</name>
-            <user>crims0n</user>
-            <email>vivek@razorpay.com</email>
+            <name>Mayank Amencherla</name>
```
Comment (captn3m0) [inline package.xml]: You can have multiple authors instead?

File: app/code/community/Razorpay/Payments/Helper/Data.php [hunk 2]
```diff
@@ -46,14 +46,19 @@ public function createOrder($order)
...
-            'currency'  => $currency
+            'receipt'         => $orderId,
+            'amount'          => $amount,
+            'currency'        => $currency,
+            'payment_capture' => 1
         );
 
         $response = $this->sendRequest($url, 'POST', $data);
 
+        session_start();
```
Comment (captn3m0) [inline app/code/community/Razorpay/Payments/Helper/Data.php]: Don't use native PHP sessions.

File: package.xml (hunk 1 above)
Comment (mayankamencherla) [inline package.xml]: Yes agreed, @73sl4 insisted I have only my name on there! So let's add your name too, vivek?

File: package.xml [hunk 3]
```diff
@@ -1,85 +1,91 @@
...
+    <stability>stable</stability>
+    <license>MIT</license>
+    <channel>community</channel>
+    <extends/>
+    <summary>Razorpay extension to accept payments from multiple banks and wallets.</summary>
+    <description>Razorpay is a payment aggregator to accept payments from various method including netbanking, credit card, debit card, wallets etc.</description>
+    <notes>This version uses the Razorpay API v1 for the best possible experience.</notes>
+    <authors>
+        <author>
+            <name>Abhay Rana</name>
```
Comment (captn3m0) [inline package.xml]: @mayankamencherla This should be multiple author tags, I think

Comment (mayankamencherla) [inline package.xml]: @captn3m0 fixed

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 4]
```diff
@@ -110,55 +110,16 @@ public function captureOrder(Varien_Object $order, $amount)
...
-            }
-            else
-            {
-                $desc = 'Payment failed. Most probably user closed the popup.';
-            }
-
-            $order->setState(Mage_Sales_Model_Order::STATE_CANCELED, true);
-            $order->addStatusHistoryComment($desc);
-            $order->save();
+        if($signature === $response['razorpay_signature']){
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: Use `hash_equal()`?

Comment (mayankamencherla) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: Done @73sl4

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 5]
```diff
@@ -110,55 +110,16 @@ public function captureOrder(Varien_Object $order, $amount)
...
     {
-        $order = Mage::getModel('sales/order');
-        $payment = Mage::getModel('sales/order_payment');
-        $order->loadByIncrementId($response['magento_order_id']);
-
-        $payment->setOrder($order);
-
+        // I have to use the order_id from the server and not the response object.
+        
+        $signature = hash_hmac('sha256', Mage::getSingleton('core/session')->getRazorpayOrderID() . "|" . $response['razorpay_payment_id'], $this->getConfigData('key_secret'));
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: split it in multiple lines.

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 6]
```diff
@@ -110,55 +110,16 @@ public function captureOrder(Varien_Object $order, $amount)
...
-            }
-            else
-            {
-                $desc = 'Payment failed. Most probably user closed the popup.';
-            }
-
-            $order->setState(Mage_Sales_Model_Order::STATE_CANCELED, true);
-            $order->addStatusHistoryComment($desc);
-            $order->save();
+        if( hash_equals($signature , $response['razorpay_signature']) ){
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: space after if and opening bracket should go in next line

**Comment by 73sl4** [comment]
@mayankamencherla LGTM. Want to test it before merging though.

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php (hunk 6 above)
Comment (mayankamencherla) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: Will fix and push shortly @73sl4

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 7]
```diff
@@ -110,55 +110,17 @@ public function captureOrder(Varien_Object $order, $amount)
...
-            catch (\Exception $e)
-            {
-                $success = false;
-            }
-        }
-
-        if ($success === true)
+        if (hash_equals($signature , $response['razorpay_signature']))
         {
-            $order->sendNewOrderEmail();
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: @mayankamencherla please add this part back. We need to store the Razorpay payment id in the order logs for reference. Everything else is fine.

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 8]
```diff
@@ -110,54 +110,34 @@ public function captureOrder(Varien_Object $order, $amount)
...
+    public function validateSignature($response)
     {
-        $order = Mage::getModel('sales/order');
-        $payment = Mage::getModel('sales/order_payment');
-        $order->loadByIncrementId($response['magento_order_id']);
-
-        $payment->setOrder($order);
-
-        $success = false;
+        // I have to use the order_id from the server and not the response object.
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: Remove this.

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 9]
```diff
@@ -110,54 +110,34 @@ public function captureOrder(Varien_Object $order, $amount)
...
-        {
-            $paymentId = $response['razorpay_payment_id'];
+        $requestFields = Mage::app()->getRequest()->getPost();
 
-            try
-            {
-                $amount = $order->getBaseGrandTotal();
+        $paymentId = $requestFields['razorpay_payment_id'];
+        
+        $signature = hash_hmac('sha256', Mage::getSingleton('core/session')->getRazorpayOrderID() . "|" . $response['razorpay_payment_id'], $this->getConfigData('key_secret'));
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: split it in multiple lines.

File: app/code/community/Razorpay/Payments/Model/Paymentmethod.php [hunk 10]
```diff
@@ -110,54 +110,34 @@ public function captureOrder(Varien_Object $order, $amount)
...
-        if ($success === true)
+        if (hash_equals($signature , $response['razorpay_signature']))
         {
+            $success = true;
             $order->sendNewOrderEmail();
             $order->setState(Mage_Sales_Model_Order::STATE_PROCESSING, true);
             $order->addStatusHistoryComment('Payment Successful. Razorpay Payment Id:'.$paymentId);
             $order->save();
         }
+
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/Model/Paymentmethod.php]: remove extra line break.

File: app/code/community/Razorpay/Payments/controllers/CheckoutController.php [hunk 11]
```diff
@@ -44,7 +44,9 @@ public function successAction()
 
         $model = Mage::getModel('razorpay_payments/paymentmethod');
 
-        $success = $model->capturePayment($response);
+        $success = $model->validateSignature($response); 
+
+        Mage::getSingleton('core/session')->unsRazorpayOrderID();
```
Comment (73sl4) [inline app/code/community/Razorpay/Payments/controllers/CheckoutController.php]: Add a comment here.

**Pull request merged**